### DISCUSSION TITLE
Totally confused with how I'm misusing StateProvider in build method

### DISCUSSION THREAD

**Discussion by mark8044** [rrousselGit/riverpod#1224] [Q&A] [answered]
What I have is a `ListView` which programmatically scrolls to a certain item by doing:
```
Scrollable.ensureVisible(_scrollToThisOneKey.currentContext!,alignment: 0, duration: const Duration(seconds: 1), curve: Curves.easeIn);
```

Now the catch is, I want it to trigger the scroll after all the images on the page have loaded, so I figured providers are the way to do that.  I created the following simple provider:

```
final testProvider = StateProvider((ref) => 0);
```

Then In my ImageWidget, I will set the state to `100` when the image loads.  Im setting it to 100 for now as an example, I plan to actually complicate this by putting in the actual image height into there, but for now think of 100 = `true` and 0 = `false`.  True being the image is loaded and False being no images loaded.

```
class _MyImageState3 extends ConsumerState<MyImage3> {
  @override
  Widget build(BuildContext context) {

    return CachedNetworkImage(
      imageUrl: widget.input,
      imageBuilder: (context, imageProvider) {
        debugPrint('- Image is loaded from network');
        ref.watch(testProvider.state).state = 100;

        return Image(image: imageProvider);
      },
    );
  }
}
```

And then in the Widget that builds my `ListView` I have the following in the `build` method to actually read the value, and if the image has been loaded, then it will do the listview Scroll:

```
      if (ref.watch(testProvider.state).state == 100)
      {
          debugPrint('- Image is loaded, lets scroll');
          ref.watch(testProvider.state).state = 0; // Im setting it back to 0 to say that the scroll was done and if another image loads, then it will re-scroll
           Scrollable.ensureVisible(_scrollToThisOneKey.currentContext!,alignment: 0, duration: const Duration(seconds: 1), curve: Curves.easeIn);
      }
```

### Now at this point I get a bunch of setState errors ###
```
setState() or markNeedsBuild() called during build.
This UncontrolledProviderScope widget cannot be marked as needing to build because the framework is already in the process of building widgets. A widget can be marked ...
```

Ok fair enough.  So I proceed to wrap my ImageWidget call in `addPostFrameCallback`

```
class _MyImageState3 extends ConsumerState<MyImage3> {
  @override
  Widget build(BuildContext context) {

    return CachedNetworkImage(
      imageUrl: widget.input,
      imageBuilder: (context, imageProvider) {
       WidgetsBinding.instance?.addPostFrameCallback((_) {
             debugPrint('- Image is loaded from network');
             ref.watch(testProvider.state).state = 100;
       });

        return Image(image: imageProvider);
      },
    );
  }
}
```

That moves the ball forward slightly, but Im still getting setState errors, so I do the same inside my `ListView` build method:

```
      if (ref.watch(testProvider.state).state == 100)
        {

          WidgetsBinding.instance?.addPostFrameCallback((_) {
            debugPrint('- Image is loaded, lets scroll');
          ref.watch(testProvider.state).state = 0; // Im setting it back to 0 to say that the scroll was done and if another image loads, then it will re-scroll
           Scrollable.ensureVisible(_scrollToThisOneKey.currentContext!,alignment: 0, duration: const Duration(seconds: 1), curve: Curves.easeIn);
          });
        }
```

### BUT ###
But, now I'm in an endless loop of:
```
- Image is loaded, lets scroll
- Image is loaded from network
- Image is loaded, lets scroll
- Image is loaded from network
- Image is loaded, lets scroll
- Image is loaded from network
- Image is loaded, lets scroll
- Image is loaded from network
- Image is loaded, lets scroll
- Image is loaded from network
- Image is loaded, lets scroll
- Image is loaded from network
```

That just keeps looping over and over in my console, the `.watch` keeps triggering for some reason.  And this is constantly triggering my `ensureVisible` scroll, so the listview actually becomes unscrollable by the user.  Whats also interesting is that it looks like the debugPrint in the ListView build is actually called first? Strange, it seems backwards from what I would have expected

For what I'm trying to accomplish, this is all just a proof of concept.  Im trying to see if I can successfully tell my ListView when an individual image has been loaded and so it can go ahead and do a scroll (or whatever else I need it to do down the line).   

Any idea how I'm misusing these providers?   The endless looping is `throwing me for a loop`...  I haven't had any problem like this before with providers and thought I knew how to use them until now

**Comment by rrousselGit** [comment, marked as answer]
Modifying providers within the build method is not allowed.

`imageBuilder` is not the correct place for such logic. It's likely that `CachedNetworkImage` exposes a different life-cycle for when the image is loaded instead of when the widget is building

**Reply by mark8044** [reply]
I believe `imageBuilder` is the only place where I can have any logic before the image is shown.   

Do you think using providers to try to do this is altogether incorrect?

On a side note, I changed the ListView build method to use `.notifier` instead of `.state` and the endless repeating loop is solved, but the order is still backwards (Listview `.watch` completes first for some reason) and so it remains non functional

```
      if (ref.watch(testProvider.notifier).state >0)
        {

          WidgetsBinding.instance?.addPostFrameCallback((_) {
            debugPrint('Image is loaded, lets scroll');
            ref.watch(testProvider.notifier).state = 0;

            Scrollable.ensureVisible(...);

          });
        }
```

**Reply by rrousselGit** [reply]
> I believe imageBuilder is the only place where I can have any logic before the image is shown.


No, this is never going to be the correct place.

If the widget does not offer another life-cycle, you should raise an issue to ask for one. You shouldn't use `imageBuilder` here

**Reply by mark8044** [reply]
Ok got it, thank you
